di/main.py

```diff
@@ -456,3 +458,146 @@ def __get__(self, inst, cls):
             return self.dependencies[self.class_obj]
         except KeyError:
             raise LookupError('Unable to find an instance for {0}'.format(self.class_obj))
+
+
+class ProxyDependencyMap(DependencyMap):
+    """
+    Derivated class that implements the proper __call__ method
+    to return an InjectorProxy .
+    """
+    def __call__(self, key):
+        """ Proxy factory method.
+            >>> dm = ProxyDependencyMap()
+            >>> my_injected_dep = dm(Spam)
+        """
+        return InjectorProxy(key, self)
+
+
+class InjectorProxy(object):
+    """
+    Alternate way of using the injector with a Proxy
+
+        >>> dm = ProxyDependencyMap()
+        >>> myfoo = dm(FOO)
+
+    This code is based on the LocalProxy implemented by the Werkzeug
+    https://github.com/pallets/werkzeug/blob/master/werkzeug/local.py#L254
+    """
+
+    def __init__(self, class_obj, dependencies):
+        object.__setattr__(self, '_InjectorProxy__class_obj', class_obj)
+        object.__setattr__(self, '_InjectorProxy__dependencies', dependencies)
+
+    def _get_current_object(self):
+        try:
+            return self.__dependencies[self.__class_obj]
+        except KeyError:
+            raise LookupError('Unable to find an instance for {0}'.format(self.class_obj))
+
+    @property
+    def __dict__(self):
+        try:
+            return self._get_current_object().__dict__
+        except RuntimeError:
```

Comment:
I think this `RuntimeError` is an implementation detail from the original code, if the current object is not defined we'll raise a `LookupError`, if the dependency is mapped to something without a `__dict__` attribute it'll raise the proper `AttributeError` already.

IMHO we should allow `LookupError` to bubble up to the calling site since otherwise it's going to be hard to debug what's going on with the proxied dependency.